sematic/graph.py

```diff
@@ -0,0 +1,560 @@
+
+        """
+        value_by_artifact_id: Dict[str, Any] = {}
+
+        futures_by_original_id: Dict[RunID, AbstractFuture] = {}
+        input_artifacts: Dict[RunID, Dict[str, Artifact]] = defaultdict(dict)
+        output_artifacts: Dict[RunID, Artifact] = {}
+
+        skip_run_ids, reset_run_ids = (
+            ([], []) if reset_from is None else self._get_skip_reset_run_ids(reset_from)
+        )
+
+        # run order guarantees parents and upstream come first
+        # This is necessary because we want upstream cloned futures
+        # to be created before downstreams so that the appropriate
+        # kwargs can be built
+        run_ids_by_execution_order = self._sorted_run_ids_by_layer(
+            run_sorter=self._execution_order
+        )
+
+        def _get_artifact_value(artifact: Artifact) -> Any:
+            if artifact.id not in value_by_artifact_id:
+                value_by_artifact_id[artifact.id] = get_artifact_value(
+                    artifact, storage
+                )
+
+            return value_by_artifact_id[artifact.id]
+
+        def _get_run_inputs(
+            run_id: RunID,
+        ) -> Tuple[Dict[str, Any], Dict[str, Artifact]]:
+            kwargs: Dict[str, Any] = {}
+
+            input_edges = self._edges_by_destination_id[run_id]
+
+            run_input_artifacts: Dict[str, Artifact] = {}
+
+            for input_edge in input_edges:
+                if input_edge.destination_name is None:
+                    raise RuntimeError("Input edge misses destination name")
+
+                value = None
+
+                if input_edge.artifact_id is not None:
+                    artifact = self._artifacts_by_id[input_edge.artifact_id]
+                    run_input_artifacts[input_edge.destination_name] = artifact
+                    value = _get_artifact_value(artifact)
+
+                if input_edge.source_run_id is not None:
+                    kwargs[input_edge.destination_name] = futures_by_original_id[
+                        input_edge.source_run_id
+                    ]
+                elif input_edge.artifact_id is not None:
+                    kwargs[input_edge.destination_name] = value
+                else:
+                    raise RuntimeError(
+                        "Invalid input edge had no source run or associated artifact"
+                    )
+
+            return kwargs, run_input_artifacts
+
+        def _get_run_output(run_id: RunID) -> Tuple[Any, Optional[Artifact]]:
+            output_edges = self._edges_by_source_id[run_id]
+
+            for output_edge in output_edges:
+                if output_edge.artifact_id is None:
+                    return None, None
+
+                artifact = self._artifacts_by_id[output_edge.artifact_id]
+                value = _get_artifact_value(artifact)
+                return value, artifact
+
+            return None, None
+
+        def _set_parent_nested(future: AbstractFuture, run: Run):
```

Comment:
I feel like this might benefit from a very brief docstring; it's not immediately obvious from the name what it does and you have to read a bit to undertstand what it's doing.